### PR TITLE
Sandbox Expenses Page Integration

### PR DESCRIPTION
**Pull request by hugo-mp3**
---
name: Expenses Page (Sandbox) 🚀
about: Flesh out expenses page with desired features 
title: "Added/Changed several features of the expenses page 🌟"
labels: ''
assignees: ''

---

**Related Issue(s)** 🔗
Please link the issue(s) this pull request addresses.

**Acceptance Test(s)** ✔️
cd into Web, then run npm run dev.
cd into Server, then run npm start.
If you haven't linked any accounts with plaid, do so. 
After you have linked an account head over to expenses page and you should be able to see your "transactions". This is only set up to use sandbox env so they will be fake. 
From there, try playing around with filtering as well as sorting. 

**Implementation Notes** 📝
BACKEND:

Server/routes/api/addExpense.js:
- takes important transaction information the user provides on the frontend to insert a transaction into the transaction table in supabase

Server/routes/api/createBankAccount.js:
- takes important account information the user provides on the frontend to insert a new account into the accounts table in supabase
- checks if an existing account with the same name exists in the accounts table prior to inserting (will not create new account if this is the case)

Server/routes/api/createUserToken.js:
- basically nothing, just one logging line I believe while I was looking for an error

Server/routes/api/deleteExpense.js:
- takes the transaction id to be deleted from the frontend and deletes it from the transactions table

Server/routes/api/getExpenses.js:
- removed the destructuring of the "limit" variable from the request body. Previously this was a value passed from the frontend to indicate how many transactions to be pulled from supabase/plaid and show to the user. (this became obsolete due to pagination implementation)
- handled access_token(s) under user for plaid expiring by implementing a system to refresh them.  (under this line: if (selectError.code === 'PGRST301'), I made a few comments that I made months ago to remember that some logic might need to be changed but I'm now unsure of their validity. I don't believe anything need be changed with the current implementation as it should avoid the edge case I foresaw back then. But I will revisit this concern the next time I come back to work on the expenses page, which will probably be to move from sandbox to development with plaid to test with real deal data.)
------------------ if there is an existing user with access tokens and transaction history in transactions table -----------------------
- gets item ids (which are unique to different institutions linked with plaid) to then use in a query on the transactions table to ensure there is at least one transaction matching each item id. If this is not the case, then transactions will again be pulled from plaid as there is surely some missing. 
- gets transactions from transaction table if the condition above is met ^^^
- gets account information from the accounts table to pass to the frontend
-------------------------- if there is an existing user with access tokens and NO transaction history ----------------------------------
- currently pulls ALL transactions under a particular access_token, for all access tokens in the definitive token list, which is determined and ironed out before pulling from plaid (IMPORTANT NOTE: there is a cursor that is returned from these api calls, which is intended to be stored so that you only pull from the last time you pulled so you don't make the calls more expensive more than need be by having plaid get every single transaction in history from an account instead of just the new ones since the last time you pulled. I have not done this but will definitely need to upon revisiting to move to dev from sandbox)
- each iteration of the loop through the access tokens gets item ids from plaidStuff table matching that particular access token for use in inserting alongside each transaction that was pulled for that token
- as implied above ^^^ transactions are inserted into the transactions table in supabase with all their information along with the matching item id, which is important when it comes to the logic mentioned before where the transactions table is checked to ensure at least one transaction per item id under particular user. They are also inserted with a variable called plaid set to true, to indicate that these transactions came from/through plaid. This is done due to the implementation of the feature of adding transactions manually without linking plaid. (useful for those scared of plaid or cash transactions)
- account information is also inserted in the accounts table, alongside a variable also called plaid for the exact same reason above ^^^

Server/routes/api/getTxnStats.js:
- calculates transaction stats (largest transaction information, total expenses, avg daily spend, date range)
- receives all visible transactions from the frontend to calculate details from

Server/routes/api/setAccessToken.js:
- now includes institution id as part of the request body payload (I forget exactly why but I believe this could come in handy for certain features later down the line by allowing the backend to associate requests with a specific financial institution)
- added logic to fetch user token under user to ensure all associated tokens are correctly linked
- if this is the users first access token/ account linked with plaid, it will opt to update the existing row, any subsequent additions will instead insert new rows in the plaidStuff table
- currently passes the access tokens to the frontend, which absolutely HAS to be changed at some point for safety reasons. It's only done now to show on the frontend that it indeed worked, but this is irrelevant as there is sufficient logging on the backend and you can just check the database as well. Upon revision I will change this. 

Server/server.js:
- simply added routes for the new endpoints

FRONTEND:

Web/src/App.jsx  removed for Web/src/App.tsx

Web/src/App.tsx:
- added an ensure valid session function to ensure the supabase access token refreshes upon expiring on the frontend (NOTE: other changes I made have made this unnecessary to do. I will surely want to delete this upon revision it is useless.)
- uses theme context now
- changed isUserToken flow to be true since I realized it always will be for this apps use I believe, at least for now. So I will likely want to change the logic to simply not use this variable and execute without checking its value. 
- Not only wraps with chakra provider now but also the MUI provider

Web/src/ColorModeSwitcher.jsx:
- uses theme context now so MUI elements know whether its light or dark mode just the same as chakra elements

Web/src/components/AddExpenseDialog.jsx:
- Dialog for the user to input information about the transaction they are trying to add
- Date picker to select the date of transaction
- text field to input the amount
- ^^^ same for description
- ^^^ same for category
- checkbox to indicate whether this transaction is cash or not. If this is pressed, and it is a cash transaction, then the information regarding selecting an account or creating a new account disappears as it should be unnecessary at that point. 
- select for accounts to add the transaction under. This only includes accounts the user has created themselves, NOT the ones they have linked with plaid. Because those transactions should automatically update so adding your own transactions to them is discouraged and disallowed for consistency in data. 
- a link to the page for creating an account in the case that they want to create a new one or they haven't created any 
- buttons to cancel or add the transaction

Web/src/components/CustomDateRangePicker.jsx:
- start date picker
- end date picker
- shortcuts for certain date ranges
- honestly don't think I need to go into more detail and I fr can't be fucked this shit is getting so long mang

Web/src/components/CustomTablePagination.jsx:
- standard pagination with slight ui customizations
- has a button on the right to add custom value for pagination, this is saved in local storage, (edge case where they try to add an existing value is taken into consideration and disallowed)

Web/src/components/PlaidLink.tsx:
- negligible changes I don't even feel like checking this has no bearing on anything related to expenses page

Web/src/components/RoutesWrapper.jsx:
- added route to the create account page

Web/src/components/TransactionFilterDrawer.jsx:
- filter drawer for filtering through transactions
- customized and made ui responsive
- has stuff for amount filtering, date filtering, account filtering, and category filtering
- also has apply and reset buttons

Web/src/components/TransactionsTable.jsx:
- built with MUI
- uses theme context to determine light or dark mode for mui components
- creates a transaction card component for details like total expenses
- component itself takes a lot of props it either uses or continues to pass down to external components it contains
- has the 3 cards above the table for seeing the transactions stats for the visible transactions 
- sorting capabilities (stored in local storage)
- filtering capabilities
- under certain width switches to cards for txns instead of a table
- delete icon to the right of transactions that were added by the user themselves, this does not exist for transactions pulled by plaid which is intended. 

Web/src/pages/Expenses.jsx:
- has all the logic and functions that get passed down to the transactions table
- logic for filtering and sorting
- logic for adding expenses
- logic for deleting expenses
- logic for calculating visible rows (for pagination)

Web/src/supabaseClient.js:
- implemented auto refresh of supabase access tokens
- also used env variables for sensative data

Web/src/muiTheme.js:
- mui theme customization including breakpoints for responsive ui

**Related PR(s)** 🔄
List any related pull requests.

**Checklist:** ✅
- [✅ ] My code follows the style guidelines of this project
- [ ✅] I have performed a self-review of my own code
- [ ✅] I have commented my code, particularly in hard-to-understand areas
- [ ✅] I have made corresponding changes to the documentation
- [ ✅] My changes generate no new warnings
- [ ✅] I have added tests that prove my fix is effective or that my feature works
- [ ✅] New and existing unit tests pass locally with my changes
- [ ✅] Any dependent changes have been merged and published in downstream modules

**Screenshots** 📸
Include any screenshots or GIFs if applicable.

**Additional Context** ➕
Add any other context about the pull request here.

## Release Notes

- **New Features**
	- Added expense management functionality with the ability to add, delete, and view transactions.
	- Introduced bank account creation feature.
	- Enhanced transaction filtering and sorting capabilities.
	- Implemented custom date range picker and table pagination.
	- Added theme switching between light and dark modes.
	- Introduced new components for managing expenses and transactions, including dialogs and tables.

- **Improvements**
	- Upgraded Material UI components and theming.
	- Improved session management and authentication flows.
	- Enhanced error handling across server routes.
	- Optimized state management using context providers.

- **Dependencies**
	- Updated and added several npm packages, including Material UI components and date utilities.
	- Adjusted emotion and react-icons package versions.

- **Performance**
	- Streamlined API routes for expenses, bank accounts, and transaction statistics.
	- Implemented more robust error handling in server-side routes.

### CODE REVIEW

File: Server/routes/api/addExpense.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                category: [expenseData.category],
+                category_id: null,
+                pending: null,
+                payment_channel: null,
+                merchant_name: null,
+                transaction_code: null,
+                location: null,
+                currency: null,
+                iso_currency_code: 'USD',
+                unofficial_currency_code: null,
```
Comment (dangphung4) [inline Server/routes/api/addExpense.js:32]: iirc supabase should allow you to redact fields and they would be populated as null in db , having a bunch of nulls and hardcoding will be a pain in the ass for the future or if you need to make changes to business logic, etc.

Postgresql as long as they are not required fields, they would be populated as null regardless

File: Server/routes/api/addExpense.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                website: null,
+                plaid: false
+            });
+
+            if (insertTransactionError){
+                console.error('this error occurred trying to insert new transaction for user: ', insertTransactionError);
+                return response.status(500).json({ message: 'this error occurred trying to insert new transaction for user' })
+            };
+
+            console.log('successfully added expense');
```
Comment (dangphung4) [inline Server/routes/api/addExpense.js:55]: when you move to prod, avoid using prints to console, this is where telemetry and error loggings come in

File: Server/routes/api/createBankAccount.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
+const express = require('express')
+const router = express.Router();
+const { getSupabaseClientWithAuth } = require('../../services/supabase')
+const { v4: uuidv4 } = require('uuid');
+
+router.post('/', function (req,response,next){
+    Promise.resolve()
+        .then(async function() {
+            const { name, type, subtype, sbAccessToken, userID } = req.body;
+            const supabase = getSupabaseClientWithAuth(sbAccessToken);
```
[the remark this answers is not in the document]
Comment (dangphung4) [reply Server/routes/api/createBankAccount.js:10]: agreed here, need ways to validate data to avoid one off errors or difficult bugs to solve later

File: Server/routes/api/createBankAccount.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            if (existingAccountError){
+                console.error('this error occurred trying to find existing account when creating a new bank account for user: ', existingAccountError);
+                return response.status(500).json({ message: 'error occurred trying to find existing account when creating a new bank account for user' })
+            };
+
+            if (existingAccount){
+                console.log('existing account found while trying to create new bank account for user: ', existingAccount);
+                return response.status(500).json({ message: 'Existing account with same name found.' })
+            }
+
```
[the remark this answers is not in the document]
Comment (dangphung4) [reply Server/routes/api/createBankAccount.js:28]: Agreed here, one thing about race conditions is that they are notorious for being almost impossible to trace down and debug, generally when building backends and rest apis, you want to follow CQRS pattern

File: Server/routes/api/createUserToken.js [hunk 5]
```diff
@@ -13,18 +13,18 @@ router.post('/', function (req, response, next) {
         const { userID, sbAccessToken } = req.body; // Destructure `accessToken` from the request body
             
         if (!sbAccessToken) {
+            console.log('no access token provided to create user token, its required.')
             return response.status(400).json({ message: 'Access token is required.' });
         }
-
+
```
[the remark this answers is not in the document]
Comment (dangphung4) [reply Server/routes/api/createUserToken.js:19]: Didn't see that code rabbit put a comment which i put earlier, but once again generally costly and expensive to have prints to console in prod.

You can look into debugging tools or even telemetry applications in conjunction with the react app, can look into OpenTelemetry or even .NET Aspire

File: Server/routes/api/createUserToken.js [hunk 6]
```diff
@@ -47,27 +48,32 @@ router.post('/', function (req, response, next) {
...
-            return response.status(500).json({ message: 'Database error while upserting user token.' });
+          if (upsertError) {
+              // Handle Supabase upsert errors
+              console.error('Error upserting user token into database:', upsertError);
+              return response.status(500).json({ message: 'Database error while upserting user token.' });
+          }
+          response.json(user.data);
+        }catch (error){
+          console.log("Error trying to create user: ", error);
+          return response.status(500).json({ message: 'Database error while upserting user token.' });
```
[the remark this answers is not in the document]
Comment (dangphung4) [reply Server/routes/api/createUserToken.js:75]: When designing rest apis, the primary focus and strength of them is being able to give descriptive feedback on codes, such as 200s, 400s, 500s. Be sure to add this as a commit later.

File: Server/routes/api/deleteExpense.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
+const express = require('express')
+const router = express.Router();
+const { getSupabaseClientWithAuth } = require('../../services/supabase')
+const { v4: uuidv4 } = require('uuid');
+
+router.post('/', function (req,response,next){
+    Promise.resolve()
+        .then(async function() {
+            const { transaction_id, sbAccessToken, userID } = req.body;
+            const supabase = getSupabaseClientWithAuth(sbAccessToken);
```
[the remark this answers is not in the document]
Comment (dangphung4) [reply Server/routes/api/deleteExpense.js:10]: Should add some form of authentication to calls like this, especially in a security perspective

File: Server/routes/api/setAccessToken.js [hunk 8]
```diff
@@ -13,66 +14,112 @@ let ITEM_ID = null;
...
+        const { public_token, sbAccessToken, userID, institution_id} = req.body;
         if (!sbAccessToken) {
           return res.status(400).json({ message: 'Access token is required.' });
         }
 
         const supabase = getSupabaseClientWithAuth(sbAccessToken);
 
-        // Step 1: Check if a user token already exists in the database
+        // Step 1: Check if an access token and item id already exists in the database
         const { data: existingUser, error: selectError } = await supabase
```
Comment (dangphung4) [inline Server/routes/api/setAccessToken.js:25]: When you method chain like this, be sure to add indentation for readability, readable code goes a long way

File: Web/src/components/AddExpenseDialog.jsx [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+            account: event.target.checked ? '' : formData.account
+        });
+    };
+
+    const handleSubmit = () => {
+        // Validate form data
+        if ((!formData.isCash && (!formData.date || !formData.description || !formData.category || !formData.account || !formData.amount)) || (formData.isCash && (!formData.date || !formData.description || !formData.category || !formData.amount))) {
+            alert('Please fill in all fields');
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (dangphung4) [reply Web/src/components/AddExpenseDialog.jsx:73]: This suggestion is much better, if you have to chain operations like that it messes up code readability and makes future changes or refactoring a pain in the ass

Anything above 2 items should generally be put into something like a list, enum, etc

File: Web/src/components/RoutesWrapper.jsx [hunk 10]
```diff
@@ -74,6 +75,7 @@ function RoutesWrapper({ setIsAuthPath }) {
       <Route path="/forgot-password" element={<ForgotPassword />} />
       <Route path="/reset-password" element={<ResetPassword />} />
       <Route path="/link-plaid" element={<PlaidLinkPage />} />
+      <Route path="/create-bank-account" element={<CreateBankAccount />} />
```
[the remark this answers is not in the document]
Comment (dangphung4) [reply Web/src/components/RoutesWrapper.jsx:78]: Generally only routes that should not be authenticated are pages/routes that don't require a user to be signed in, everything else should require some form of authentication for security

File: Web/src/pages/CreateBankAccount.jsx [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+                    'Content-Type': 'application/json',
+                },
+                body: JSON.stringify({
+                    name: accountName,
+                    type: accountType,
+                    subtype: accountSubtype,
+                    sbAccessToken: sbAccessToken,
+                    userID: userID,
+                }),
+            });
```
[the remark this answers is not in the document]
Comment (dangphung4) [reply Web/src/pages/CreateBankAccount.jsx:66]: Agreed here, use a .env.local or .env when you can for everything, you will forget about this in the future and run into errors like I did because of the URL

**Review by dangphung4 — CHANGES_REQUESTED** [review]

**Comment by dangphung4** [comment]
Review still WIP, a lot of code to look at here
Good work!

nit: in the future if we can make prs ideally 500 lines of code or less that would be better (smaller reviews more often tends to lead to faster development)

**Comment by dangphung4** [comment]
> Review still WIP, a lot of code to look at here
> 
> Good work!
> 
> 
> 
> nit: in the future if we can make prs ideally 500 lines of code or less that would be better (smaller reviews more often tends to lead to faster development)

Also add screenshots in pr when you can, and keep PR readme small and lean as possible. The small things like html do not matter, just need major changes

**Comment by dangphung4** [comment]
I should be able to know what you did defined in your commit history, rather than you putting every little thing you did on the pr readme, each commit should be a feature or a set of changes that are retracible and reversible

File: Server/routes/api/setAccessToken.js [hunk 12]
```diff
@@ -13,66 +14,112 @@ let ITEM_ID = null;
...
+                  item_id: ITEM_ID,
+              })
+              .eq('id', userID); // Update based on userID
+
+          if (updateError) {
+              console.error('Error updating access token, item id, institution id, and user token data:', updateError);
+              return res.status(500).json({ message: 'Error updating access token, item id, institution id, and user token data.' });
+          }
+
+          console.log('access token, item id, institution id, and user token data successfully updated for user:', userID);
```
[the remark this answers is not in the document]
Comment (hugo-mp3) [reply Server/routes/api/setAccessToken.js:119]: understandable, I misworded my comment but the logic is solid at least in conjunction with the rest of the plaid linking code. this logic will only ever execute after a row has been created for the user in the database, so the edge case where there is no row at all and we try to update that non existent row will never happen. I get how this was confusing but the point was really if no row matching the criteria for a non null access_token was found then update that row that definitely exists to have a new access_token so that it has a value. I changed the comment to hopefully be more clear but making this suggested change to the logic would actually break the current functionality.

File: Server/routes/api/getExpenses.js [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+                        console.log("successfully updated/refreshed expired access token")
+                        definitiveTokenList = refreshedTokenList.map(({ newToken }) => newToken);
+                    }else{
+                        console.log('JWT expired error occurred somehow while selecting access_tokens but there were no access_tokens in the database to refresh')
+                    }
+                }else{
+                    console.error('Error checking for existing user and access token to get expenses', selectError)
+                    return response.status(500).json({ message: 'Database error while while checking for access token to get expenses' })
+                }
+            }
```
[the remark this answers is not in the document]
Comment (hugo-mp3) [reply Server/routes/api/getExpenses.js:78]: understandable again, as a result I commented indicating that this error code indeed corresponds to a jwt expiration, at least for the present moment. Also, existingUser is only accessed after it is confirmed that it exists. This is necessary for the retrieval and subsequent refreshing of jwt tokens in the event of an expiration.

File: Server/routes/api/getExpenses.js [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+            allAdded = allAdded.map(transaction => ({
+                ...transaction,
+                date: transaction.date, // Standardized date format
+                plaid: true,
+            })); 
+            const recently_added = [...allAdded].sort(compareTxnsByDateAscending)
+            
+            response.json({ latest_transactions: recently_added, accounts: allAccounts });
```
[the remark this answers is not in the document]
Comment (hugo-mp3) [reply Server/routes/api/getExpenses.js:277]: in the future will meet this along with other safety concerns I have already noted but this is just sandbox information for now so no real worries or need to do it this second.

File: Web/src/components/TransactionFilterDrawer.jsx [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+            <MUIGrid container spacing={3} sx={{ mb: 2, pr: 4 }}>
+                {/* Amount Filter */}
+                <MUIGrid item xs={12} md_1={3}>
+                    <MUITypography variant="h6" gutterBottom textAlign={'center'}>Amount Range</MUITypography>
+                    <MUIBox display="flex" gap={2}>
+                        <MUITextField 
+                            label="Min Amount" 
+                            type="number"
+                            value={amountFilter.min}
+                            onChange={(e) => setAmountFilter(prev => ({ ...prev, min: e.target.value }))}
```
[the remark this answers is not in the document]
Comment (hugo-mp3) [reply Web/src/components/TransactionFilterDrawer.jsx:76]: did this but did not add min b/c some values for transactions will be negative
